server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java

```diff
@@ -207,37 +189,84 @@ public TimeSeriesIdBuilder(@Nullable IndexRouting.ExtractFromSource.Builder rout
-                if (timeSeriesId.length() > LIMIT) {
-                    throw new IllegalArgumentException(NAME + " longer than [" + LIMIT + "] bytes [" + timeSeriesId.length() + "].");
+                for (DimensionDataHolder entry : dimensions) {
+                    out.writeBytesRef(entry.name);
+                    entry.value.writeTo(out);
                 }
-                return timeSeriesId;
+                return out.bytes();
             }
         }
 
+        /**
+         * Here we build the hash of the tsid using a similarity function so that we have a result
+         * with the following pattern:
+         *
+         * hash128(catenate(dimension field names)) +
+         * foreach(dimension field value, limit = MAX_DIMENSIONS) { hash32(dimension field value) } +
+         * hash128(catenate(dimension field values))
+         *
+         * The idea is to be able to place 'similar' time series close to each other. Two time series
+         * are considered 'similar' if they share the same dimensions (names and values).
+         */
+        public BytesReference withHash() throws IOException {
+            // NOTE: hash all dimension field names
+            int numberOfDimensions = Math.min(MAX_DIMENSIONS, dimensions.size());
+            int tsidHashIndex = 0;
+            byte[] tsidHash = new byte[16 + 16 + 4 * numberOfDimensions];
+
+            tsidHasher.reset();
+            for (final DimensionDataHolder dimension : dimensions) {
+                tsidHasher.update(dimension.name.bytes);
+            }
+            tsidHashIndex = writeHash128(tsidHasher.digestHash(), tsidHash, tsidHashIndex);
+
+            // NOTE: concatenate all dimension value hashes up to a certain number of dimensions
+            int tsidHashStartIndex = tsidHashIndex;
+            for (final DimensionDataHolder dimension : dimensions) {
+                if ((tsidHashIndex - tsidHashStartIndex) >= 4 * numberOfDimensions) break;
+                final BytesRef dimensionValueBytesRef = dimension.value.toBytesRef();
+                ByteUtils.writeIntLE(
+                    StringHelper.murmurhash3_x86_32(
+                        dimensionValueBytesRef.bytes,
+                        dimensionValueBytesRef.offset,
+                        dimensionValueBytesRef.length,
+                        SEED
+                    ),
+                    tsidHash,
+                    tsidHashIndex
+                );
+                tsidHashIndex += 4;
+            }
+
+            // NOTE: hash all dimension field allValues
+            tsidHasher.reset();
+            for (final DimensionDataHolder dimension : dimensions) {
+                tsidHasher.update(dimension.value.toBytesRef().bytes);
+            }
+            tsidHashIndex = writeHash128(tsidHasher.digestHash(), tsidHash, tsidHashIndex);
+
+            assert tsidHashIndex == tsidHash.length;
+            try (BytesStreamOutput out = new BytesStreamOutput(tsidHash.length)) {
```

Comment:
Passing `tsidHash.length` here is important...if not it might happen that a larger buffer is allocated and the additional bytes are later encoded in the Base64 encoding.